File: rgctl-tests/ecommerce-csharp/src/Ecommerce/Coolstore/CartEndpoint.cs
using Ecommerce.Coolstore.Models;
using Ecommerce.Coolstore.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Ecommerce.Coolstore;

[ApiController]
[AllowAnonymous]
[Route("services/cart")]
public class CartEndpoint : ControllerBase
{
    private readonly ShoppingCartService _shoppingCartService;

    public CartEndpoint(ShoppingCartService shoppingCartService)
    {
        _shoppingCartService = shoppingCartService;
    }

    [HttpGet("{cartId}")]
    public ActionResult<ShoppingCart> GetCart(string cartId) =>
        Ok(_shoppingCartService.GetShoppingCart(cartId));

    [HttpPost("checkout/{cartId}")]
    public ActionResult<ShoppingCart> Checkout(string cartId) =>
        Ok(_shoppingCartService.CheckOutShoppingCart(cartId));

    [HttpPost("{cartId}/{itemId}/{quantity:int}")]
    public ActionResult<ShoppingCart> Add(string cartId, string itemId, int quantity)
    {
        var cart = _shoppingCartService.GetShoppingCart(cartId);
        var product = _shoppingCartService.GetProduct(itemId);
        if (product == null)
        {
            return Ok(cart);
        }

        var sci = new ShoppingCartItem
        {
            Product = product,
            Quantity = quantity,
            Price = product.Price
        };
        cart.AddShoppingCartItem(sci);
        _shoppingCartService.PriceShoppingCart(cart);
        cart.ShoppingCartItemList = _shoppingCartService.DedupeCartItems(cart.ShoppingCartItemList);
        _shoppingCartService.PriceShoppingCart(cart);
        return Ok(cart);
    }

    [HttpDelete("{cartId}/{itemId}/{quantity:int}")]
    public ActionResult<ShoppingCart> Delete(string cartId, string itemId, int quantity)
    {
        var cart = _shoppingCartService.GetShoppingCart(cartId);
        var toRemove = new List<ShoppingCartItem>();
        foreach (var sci in cart.ShoppingCartItemList)
        {
            if (sci.Product != null && itemId == sci.Product.ItemId)
            {
                if (quantity >= sci.Quantity)
                {
                    toRemove.Add(sci);
                }
                else
                {
                    sci.Quantity -= quantity;
                }
            }
        }

        foreach (var sci in toRemove)
        {
            cart.RemoveShoppingCartItem(sci);
        }

        _shoppingCartService.PriceShoppingCart(cart);
        return Ok(cart);
    }
}

File: rgctl-tests/ecommerce-java/src/main/java/com/example/ecommerce/coolstore/rest/CartEndpoint.java
package com.example.ecommerce.coolstore.rest;

import com.example.ecommerce.coolstore.model.CatalogProduct;
import com.example.ecommerce.coolstore.model.ShoppingCart;
import com.example.ecommerce.coolstore.model.ShoppingCartItem;
import com.example.ecommerce.coolstore.service.ShoppingCartService;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.ArrayList;
import java.util.List;

@RestController
@RequestMapping("/services/cart")
public class CartEndpoint {

    private final ShoppingCartService shoppingCartService;

    public CartEndpoint(ShoppingCartService shoppingCartService) {
        this.shoppingCartService = shoppingCartService;
    }

    @GetMapping("/{cartId}")
    public ShoppingCart getCart(@PathVariable String cartId) {
        return shoppingCartService.getShoppingCart(cartId);
    }

    @PostMapping("/checkout/{cartId}")
    public ShoppingCart checkout(@PathVariable String cartId) {
        return shoppingCartService.checkOutShoppingCart(cartId);
    }

    @PostMapping("/{cartId}/{itemId}/{quantity}")
    public ShoppingCart add(
            @PathVariable String cartId,
            @PathVariable String itemId,
            @PathVariable int quantity) {
        ShoppingCart cart = shoppingCartService.getShoppingCart(cartId);
        CatalogProduct product = shoppingCartService.getProduct(itemId);
        if (product == null) {
            return cart;
        }
        ShoppingCartItem sci = new ShoppingCartItem();
        sci.setProduct(product);
        sci.setQuantity(quantity);
        sci.setPrice(product.getPrice());
        cart.addShoppingCartItem(sci);
        shoppingCartService.priceShoppingCart(cart);
        cart.setShoppingCartItemList(shoppingCartService.dedupeCartItems(cart.getShoppingCartItemList()));
        shoppingCartService.priceShoppingCart(cart);
        return cart;
    }

    @DeleteMapping("/{cartId}/{itemId}/{quantity}")
    public ShoppingCart delete(
            @PathVariable String cartId,
            @PathVariable String itemId,
            @PathVariable int quantity) {
        ShoppingCart cart = shoppingCartService.getShoppingCart(cartId);
        List<ShoppingCartItem> toRemove = new ArrayList<>();
        for (ShoppingCartItem sci : cart.getShoppingCartItemList()) {
            if (sci.getProduct() != null && itemId.equals(sci.getProduct().getItemId())) {
                if (quantity >= sci.getQuantity()) {
                    toRemove.add(sci);
                } else {
                    sci.setQuantity(sci.getQuantity() - quantity);
                }
            }
        }
        toRemove.forEach(cart::removeShoppingCartItem);
        shoppingCartService.priceShoppingCart(cart);
        return cart;
    }
}
